# Number Cross — Jane Street, August 2014

https://www.janestreet.com/puzzles/number-cross-index/

## Solution

**407358** — the sum of the eight 5-digit entries (18-across, 20-across, 26-across, 28-across,
4-down, 6-down, 27-down, 28-down). Solved by hand; the finished grid is at the bottom.

## Goals

I'm gonna try to solve this one by hand.  Python for some helper functions and to enumberate candidates.

## The puzzle

> Place the digits 1 thru 9 (no zeroes) in the crossword grid below so that all of the clues are
> satisfied. No digit is repeated in any one grid entry (e.g. 1-across is a grid entry), and no grid
> entry is used more than once within the puzzle.
>
> Submit the sum of the 8 five-digit grid entries in the completed puzzle.

![The puzzle grid and clues](number-cross.png)

The clue list is also in [`clues.csv`](clues.csv) with each entry's length and crossings, and
[`number-cross-printable.pdf`](number-cross-printable.pdf) is the blank grid at pencil size.

## 1. Three-digit squares

In [ ]:
three_digit_squares = [i**2 for i in range(10, 32) if len(set(str(i**2))) == 3]

no_repeating_digits = [n for n in three_digit_squares if len(set(str(n))) == 3]

print(no_repeating_digits)

[169, 196, 256, 289, 324, 361, 529, 576, 625, 729, 784, 841, 961]


The set of 3 digit squares is the most compelling starting place.  There are 11 of them and only 13 candidates.  


Candidates: [169, 196, 256, 289, 324, 361, 529, 576, 625, 729, 784, 841, 961]


1 accross and 1 down share a leading number, so they have to start with 1, 2, 3, 5, or 7.  13 accross starts with a 1 because at least 196 must be used in the grid as a square and 13 across must be less than it.  Thus, 1 down is 361 and 1 across is 324.  WE remove those from candidates for other squares.

> **[ TODO — reasoning not written down ]**
>
> The jump to "1 down is 361 and 1 across is 324" skips a step. Once 13-across is known to
> start with 1, 1-down is a square ending in 1 — which leaves 361, 841 and 961. Worth writing down
> what kills the other two, since 1-across shares 1-down's leading digit and no grid entry may be
> used twice.

7 across is 3 digit square that begins 2 other 3 digit squares moving down. That helps reduce candidates.

Candidates: [169, 196, 256, 289, 529, 576, 625, 729, 784, 841, 961]

Reduced: [169, 196, 256, 289, 529, 576, 729, 784]

12 across is also a 3 digit square.  That is super restrictive.

Reduced [529, 576]

7, 12 across, and 7,8 down become a family of 529, 576, 784, 289

> **[ TODO — reasoning not written down ]**
>
> The reduction from eight candidates to `[529, 576]`, and then to the family
> `529, 576, 784, 289` for 7-across / 12-across / 7-down / 8-down, is the biggest single step in the
> solve and has no working shown. Worth spelling out, or writing the filter in code.

## 2. Helper functions, and 9-down

In [40]:
def find_primes(low, high):
    primes = []
    for n in range(low, high + 1):
        if is_prime(n):
            primes.append(n)
    return primes


def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True


def multiples_of(x, min, max):
    least_multiple = (min + x - 1) // x * x
    for i in range(least_multiple, max + 1, x):
        yield i


def filter_for_distinct(nums):
    for num in nums:
        digits = str(num)
        if "0" in digits:
            continue
        if len(digits) == len(set(digits)):
            yield num

In [8]:
print(find_primes(940, 949))
print(find_primes(690, 699))

[941, 947]
[691]


9 down is prime, and can be led by 94 or 69.

Candidates: [941, 947, 691]

However, 14 across must have digits summing to 42 using 7 distinct digits!  This is a sodoku trick meaning we must exclude 1 and 2 form the digits of 14.  Thus, the answer is 947 which sets all the squares in that region too!

## 3. 4-across, 11-across and 5-down

Square candidates: [169, 196, 256, 625, 729, 841, 961]

Now I will look at 4 and 11 across.  4 down is restricted to digits '1,2,3,4,6' and 6 down '5,6,7,8,9'.

So the valid 4 and 11 across squares are [169, 196, 256, 625]

Leading numbers possible for 5 down are 69, 65, 62, 96, 95, 92, 56, 59, 52, 26, 29, 25.

In [12]:
for multiple in multiples_of(947, 947, 10000):
    print(multiple)

947
1894
2841
3788
4735
5682
6629
7576
8523
9470


Valid solutions for 5 down are [5682] only!  Giving us our fill of 4 across is 256 and 11 accros is 169

> **[ TODO — reasoning not written down ]**
>
> Nine multiples of 947 are printed and the next line says 5682 is the only valid one. The
> filtering — distinct nonzero digits, then the crossings with 4-across and 11-across — isn't
> recorded.

## 4. 2-down

In [13]:
def divisible_by(n, min, max):
    for i in range(min, max + 1):
        if i % n == 0:
            print(i)

In [14]:
divisible_by(11, 200, 300)

209
220
231
242
253
264
275
286
297


> **[ TODO — reasoning not written down ]**
>
> This lists every multiple of 11 in the 200s but stops there. Which candidates survive the
> no-zero and distinct-digit rules, and what did the crossings with 10-across and 13-across leave?

## 5. 36-down

36 down candidates [729, 841]

[625, 729, 961]

> **[ TODO — reasoning not written down ]**
>
> Two bare candidate lists with no note attached. What entry is the second list for, and what
> narrowed 36-down to `[729, 841]`?

In [16]:
for multiple in multiples_of(123, 123, 1000):
    print(multiple)

123
246
369
492
615
738
861
984


## 6. 39-across

In [21]:
find_primes(450, 500)

[457, 461, 463, 467, 479, 487, 491, 499]

In [22]:
974 - 162

812

39 accross is either 491 * 2 = 982 or 487 * 2 = 974

> **[ TODO — reasoning not written down ]**
>
> 39-across came out as 974 rather than 982. The tiebreak isn't recorded.

## 7. 29-across and 35-down

OK 29 across is pretty restricted by the 4 in second positions, the fact that it's a multiple of 256, and the distinct digit rule.  Lets figure out how many candidates we have.  Additionally, the second digit of 35 down must be 7 or 8, further restricting 29

In [ ]:
candidates_29 = list(filter_for_distinct(list(multiples_of(256, 100000, 256 * 987))))

candidates_29 = [candidate for candidate in candidates_29 if str(candidate)[1] == "4"]


print(candidates_29)

candidate_35 = [candidate / 256 for candidate in candidates_29]

print(candidate_35)

[143872, 147968, 148736, 243968, 248576, 249856]
[562.0, 578.0, 581.0, 953.0, 971.0, 976.0]


The only valid candidates for 35 down are 578 and 976, which makes the only valid candidates for 29 across 147968 and 249856.

This finding forces a 7 in the second positions of 35 down and 39 accross, forcing a 4 in second position of 36 down, giving us 36 down as 841.

Progress!!

## 8. 18-across and 31-down

In [62]:
print(str(69578 + 125))
print(str(69578 + 126))
print(str(69578 + 135))
print(str(69578 + 136))
print(str(69578 + 145))
print(str(69578 + 146))
print(str(69578 + 156))
print("________________")
print(str(69578 + 235))
print(str(69578 + 236))
print(str(69578 + 245))
print(str(69578 + 246))
print(str(69578 + 256))

69703
69704
69713
69714
69723
69724
69734
________________
69813
69814
69823
69824
69834


Figured out that 31 down must be 213, otherwise 37 across plus 6 down starts 698, which we can't do because the 8's are used up in 24 and 26 across positions 3 and 4.

In [63]:
print(list(multiples_of(213, 200, 1000)))

[213, 426, 639, 852]


In [ ]:
974 - 132

> **[ TODO — reasoning not written down ]**
>
> The two bare subtractions (`974 - 162`, `974 - 132`) are working for 38-across, which is
> 39-across minus 13-across. Which value of 13-across won, and why, isn't written down.

## 9. 15-down

In [71]:
candidates_15 = list(filter_for_distinct(list(multiples_of(11, 100000, 1000000))))

candidates_15 = [
    candidate
    for candidate in candidates_15
    if str(candidate)[0] == "3"
    and str(candidate)[1] == "8"
    and str(candidate)[3] == "1"
    and str(candidate)[5] == "5"
]


print(candidates_15)

[382195, 384175, 387145, 389125]


> **[ TODO — reasoning not written down ]**
>
> The filter fixes 15-down's digits in positions 1, 2, 4 and 6 as 3, 8, 1 and 5. Those came
> from crossings — worth naming which ones, since it's the step that makes the cell work.

## 10. 25-across and 42-across

In [ ]:
candidates_25 = [577 + 100 * n for n in range(10)]

candidates_25 = [candidate + 10 * n for n in [6, 8] for candidate in candidates_25]

candidates_25 = list(
    filter_for_distinct([candidate for candidate in candidates_25 if candidate > 1100])
)

print(candidates_25)

candidates_42 = [candidate - 576 for candidate in candidates_25]
print(candidates_42)

[1237, 1437, 1537, 1257, 1357, 1457]
[661, 861, 961, 681, 781, 881]


> **[ TODO — reasoning not written down ]**
>
> `577 + 100 * n`, then `+ 10 * n` for n in `[6, 8]`, then a floor of 1100 — this builds the
> 25-across candidates from 42-across + 7-down, but the shape of the construction needs a sentence to
> be readable later.

## 11. 22-down

In [ ]:
candidates_22 = find_primes(100, 1000)

print(candidates_22)

candidates_22 = [
    candidate
    for candidate in candidates_22
    if str(candidate)[2] == "7" and str(candidate)[1] in "35"
]

print(candidates_22)

[101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]
[137, 157, 257, 337, 457, 557, 757, 857, 937]


> **[ TODO — reasoning not written down ]**
>
> The filtered prime list is printed but no conclusion is recorded.

Alright and in a flurry we're done!

> **[ TODO — reasoning not written down ]**
>
> The flurry covered roughly a third of the grid with no notes: 3-down, 10-across, 16-across,
> 17-down, 20-across, 21-down, 23-down, 28-across, 28-down, 30-down, 31-across, 34-across, 37-across
> and 40-across. If this notebook is meant to be re-readable, this is the section most worth
> reconstructing while it's still fresh.

## The finished grid

![The finished grid](finished-grid.jpg)

As text — `#` is a blocked square, and this is the same grid as
[`progress.txt`](progress.txt):

```
324#256#529
659#169#784
132#4853697
##136278###
69713#87219
2658###1437
51893#34578
###249856##
2684751#798
135#842#974
392#925#861
```

Checked against all 48 clues: every entry satisfies its clue, no entry contains a zero or a
repeated digit, and no two entries are the same number.

In [ ]:
# The eight 5-digit entries, which are what the puzzle asks us to add up.
five_digit_entries = {
    "18-across": 69713,
    "20-across": 87219,
    "26-across": 51893,
    "28-across": 34578,
    "4-down": 21463,
    "6-down": 69578,
    "27-down": 34789,
    "28-down": 38125,
}

for entry, value in five_digit_entries.items():
    print(f"{entry:>10} = {value}")

print()
print("answer:", sum(five_digit_entries.values()))

## The worked clue sheet

Every clue crossed off as it fell, with the candidate lists in the margin.

![The clue list, worked](worked-clues.jpg)